src/metabase/lib/schema/expression.cljc

```diff
@@ -1,39 +1,187 @@
 (ns metabase.lib.schema.expression
   (:require
+   [malli.core :as mc]
+   [metabase.lib.schema.common :as common]
    [metabase.lib.schema.filter :as filter]
    [metabase.lib.schema.literal :as literal]
    [metabase.lib.schema.ref :as ref]
+   [metabase.lib.schema.temporal-bucketing :as temporal-bucketing]
+   [metabase.types]
    [metabase.util.malli.registry :as mr]))
 
+(comment metabase.types/keep-me)
+
+(defn- ref-of-base-type [base-type]
+  [:and
+   [:ref ::ref/ref]
+   [:fn
+    {:error/message (str ":field, :expression, :or :aggregation with a " base-type " :base-type")}
+    (fn [[_ref opts]]
+          (isa? (:base-type opts) base-type))]])
+
+(defn- ref-of-base-type-with-temporal-unit [base-type temporal-unit-schema]
+  [:and
+   (ref-of-base-type base-type)
+   [:fn
+    {:error/message (str ":field, :expression, :or :aggregation reference returning a "
+                         base-type
+                         " with a :temporal-unit that matches "
+                         ;; TODO -- humanize this
+                         (pr-str temporal-unit-schema))}
+    (let [validator (mc/validator temporal-unit-schema)]
+      (fn [[_ref {:keys [temporal-unit], :as _opts}]]
+        (validator temporal-unit)))]])
+
 ;;; An expression that we can filter on, or do case statements on, etc.
 (mr/def ::boolean
   [:or
-   ::literal/boolean
-   [:ref ::filter/filter]])
+   {:error/message "expression returning a boolean"}
+   [:ref ::literal/boolean]
+   [:ref ::filter/filter]
+   (ref-of-base-type :type/Boolean)])
 
 ;;; An expression that returns a string.
 (mr/def ::string
-  ::literal/string)
+  [:or
+   {:error/message "expression returning a string"}
+   [:ref ::literal/string]
+   (ref-of-base-type :type/Text)])
+
+;;; `:length` clause: returns an integer
+;;;
+;;; [:length options <string-expression>
+(mr/def ::length
+  [:tuple
+   [:= :length]
+   ::common/options
+   ::string])
+
+;;; a `:*` clause with all integer expression arguments returns an integer.
+(mr/def ::*.integer
```

Comment:
Added schemas for `:*` as proofs-of-concept. There are now two versions of this schema, one that returns an `::integer`, and one that returns "non-integer reals". The `::*.integer` schema is only satisfied if all of its args are `::integer` expressions as well, while the `::*.non-integer-real` version is basically happy if all of its args are numbers, and at least one of them is not an integer.

e.g.

multiplication with all integer returns an integer:

1 * 10 = 10 

multiplication with at least one non-integer real returns a non-integer real:

1 * 10.0 = 10.0

#### Multiple schemas?!

It certainly is annoying to have to have multiple schemas for each clause when we only had one with the old stuff, right? But I think the use case is compelling enough to justify it. 

1. It's not like the schemas are incredibly hard to write or understand and need to be touched all the time -- once we finish building everything out it should be fairly static.
2. Strict type checking in the schema itself will probably actually make the rest of the MLv2 project easier to do, since we'll catch invalid stuff right away in schema validation and won't have to do a bunch of manual constraint validation like we do in the QP, or wait until someone runs a query to find out it doesn't work
3. This will ultimately let us open MBQL up, for example various clauses like `:substring` take a number, and if we **knew** some expression like `[:+ {} x y]` returned an integer, we could allow it inside `:substring`. Lots of possibilities. We're already seeing some benefits related to this with the changes we've made to allow any `::boolean` expression as the condition for a join